### PR TITLE
[GPU] resolve WA of using clDNN reduce

### PR DESCRIPTION
**Pull request by byungilm**
+ Resolved layout mismatch of unreduced axes
+ Removed redundant reduce

Signed-off-by: Min, Byungil <byungil.min@intel.com>

### Details:
 - *item1*
 - *...*

### Tickets:
 - 86189

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp [hunk 1]
```diff
@@ -411,5 +429,12 @@ void prepare_buffer_fusing::run(program& p) {
             node.get_output_layout();
             node.can_be_optimized(can_reshape_be_optimized(node));
         });
+        program_helpers::do_for_types<reduce>(*node, [&p](reduce_node& node) {
+            node.get_output_layout();
+            if (can_reduce_be_optimized(node)) {
+                node.can_be_optimized(true);
+                p.extract_and_remove(node);
+            }
+        });
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp:438]: IMO, this is not related to this pass. It would be best if we can "not create" it at first place. If it is difficult not to create it, next place would be to create another pass or remove redundant reorder.

**Comment by byungilm** [comment]
This PR will be updated after resolving issues related to coming release.
1) check 3d reduction accuracy - need to check base / where dequantizes  
2) move the optimizing spot  (related to Mingyu's comment)

File: src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp (hunk 1 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp]: I created another pass for removing redundant primitives.

**Review by isanghao — COMMENTED** [review]
generally looks good to me

File: src/common/transformations/src/transformations/common_optimizations/common_optimizations.cpp [hunk 2]
```diff
@@ -84,6 +84,7 @@
 #include "transformations/op_conversions/convert_pad_to_group_conv.hpp"
 #include "transformations/op_conversions/convert_prior_box_v8_to_v0.hpp"
 #include "transformations/op_conversions/convert_reduce_to_pooling.hpp"
+#include "transformations/op_conversions/convert_reduce_to_reshape.hpp"
```
Comment (yeonbok) [inline src/common/transformations/src/transformations/common_optimizations/common_optimizations.cpp:87]: If you are going to add a new pass to common passes, you should validate for CPU too at least. 
Or you can apply it only to the gpu pass?

Comment (isanghao) [reply src/common/transformations/src/transformations/common_optimizations/common_optimizations.cpp:87]: Oh I didn't catch this. Let's add it to gpu pass only.

File: src/plugins/intel_gpu/src/plugin/ops/reduce.cpp [hunk 3]
```diff
@@ -84,30 +84,33 @@ static void CreateReduceOp(Program& p, const std::shared_ptr<ngraph::Node>& op,
 
     auto resultLayerName = layerName;
     auto out_dims = op->get_output_shape(0).size();
-    if (out_dims == 3 && !keep_dims && rank >= 4) {
+    if (!keep_dims &&
+        ((out_dims == 3 && rank >= 4) || (out_dims == 2 && rank >= 3) || (out_dims == 1 && rank >= 2))) {
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/plugin/ops/reduce.cpp:88]: ((out_dims <= 3) && (rank > out_dims + 1)) ?

File: src/common/transformations/src/transformations/common_optimizations/common_optimizations.cpp (hunk 2 above)
Comment (byungilm) [reply src/common/transformations/src/transformations/common_optimizations/common_optimizations.cpp:87]: Thanks i will removed.

File: src/common/transformations/include/transformations/op_conversions/convert_reduce_to_reshape.hpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    }
+};
+
+static bool is_redundant(ngraph::Shape input, ngraph::Shape output) {
+    if (shape_size(input) != shape_size(output))
+        return false;
+
+    for (size_t idx = 0; idx < input.size(); idx++) {
+        if (input[idx] != output[idx] && input[idx] != 1)
+            return false;
```
[the remark this answers is not in the document]
Comment (byungilm) [reply src/common/transformations/include/transformations/op_conversions/convert_reduce_to_reshape.hpp:96]: input : 7722, 1
output 7722
It is actually same tensor in reducemean
So it will be true if 1.

File: src/plugins/intel_gpu/src/plugin/ops/reduce.cpp [hunk 5]
```diff
@@ -123,10 +126,10 @@ static void CreateReduceOp(Program& p, const std::shared_ptr<ngraph::Node>& op,
                                            cldnn::reorder_mean_mode::subtract,
                                            op->get_friendly_name());
         p.AddPrimitive(reorder_prim);
-        p.AddPrimitiveToProfiler(op, reorderLayerName);
-    } else {
-        p.AddPrimitiveToProfiler(op);
+        resultLayerName = reorderLayerName;
     }
+
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/plugin/ops/reduce.cpp:131]: Please add reduce primitive to AddInnerPrimitiveToProfiler somewhre

Comment (byungilm) [reply src/plugins/intel_gpu/src/plugin/ops/reduce.cpp:131]: Fixed

File: src/plugins/intel_gpu/src/plugin/transformations_pipeline.cpp [hunk 6]
```diff
@@ -190,7 +191,22 @@ void TransformationsPipeline::apply(std::shared_ptr<ov::Model> func) {
...
+            [](const_node_ptr &node) -> bool { return true; });
+        pass_config->set_callback<ngraph::pass::ConvertReduceMaxToReshape>(
+            [](const_node_ptr &node) -> bool { return true; });
+        pass_config->set_callback<ngraph::pass::ConvertReduceMinToReshape>(
+            [](const_node_ptr &node) -> bool { return true; });
+        pass_config->set_callback<ngraph::pass::ConvertReduceLogicalAndToReshape>(
+            [](const_node_ptr &node) -> bool { return true; });
+        pass_config->set_callback<ngraph::pass::ConvertReduceLogicalOrToReshape>(
+            [](const_node_ptr &node) -> bool { return true; });
+
```
[the remark this answers is not in the document]
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/src/plugin/transformations_pipeline.cpp:202]: There's `ConvertReduceToReshape`  pass which aggregates all of the above, so I suppose it should be used instead.

File: src/common/transformations/src/transformations/op_conversions/convert_reduce_to_reshape.cpp [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+ngraph::pass::ConvertReduceLogicalOrToReshape::ConvertReduceLogicalOrToReshape() {
+    MATCHER_SCOPE(ConvertReduceLogicalOrToReshape);
+    auto m = std::make_shared<ngraph::pattern::Matcher>(
+        ngraph::pattern::wrap_type<opset1::ReduceLogicalOr>(
+            {pattern::any_input(pattern::has_static_shape()), pattern::wrap_type<opset1::Constant>()},
+            pattern::has_static_shape()),
+        matcher_name);
+    register_matcher(m, convert_reduce_to_reshape<opset1::ReduceLogicalOr>());
+}
```
Comment (isanghao) [inline src/common/transformations/src/transformations/op_conversions/convert_reduce_to_reshape.cpp:79]: nit: please add an empty line at the end

File: src/plugins/intel_gpu/src/graph/impls/onednn/reduction_onednn.cpp [hunk 8]
```diff
@@ -27,8 +27,18 @@ struct reduction_onednn : typed_primitive_onednn_impl<reduce, dnnl::reduction::d
...
-        auto output_md = onednn::layout_to_memory_desc(arg.get_output_layout());
+        auto input_layout = input.get_output_layout();
+        auto output_layout = arg.get_output_layout();
+
+        for (size_t idx = 0 ; idx < prim->axes.size(); idx++) {
+            if (output_layout.get_dim(prim->axes[idx]) != 1) {
+                reorder_unreduced_axis_no_fusion(input_layout, output_layout, prim->axes);
+                break;
+            }
+        }
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/impls/onednn/reduction_onednn.cpp:38]: * Do we need a for loop here?
* I think adding some assert would be better to understand
* The function name `reorder_unreduced_axis_no_fusion` is difficult to understand. Can you add some comment?

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp [hunk 9]
```diff
@@ -150,6 +150,21 @@ void combine_bf_with_first_spatial_dim(cldnn::layout& l) {
...
+    }
+
+    auto output_tensor = output_layout.get_tensor();
+    for (size_t idx = 0; idx < in_dims.size(); idx++) {
+        output_tensor.raw[idx] = in_dims[idx];
+    }
+
+    output_layout.set_tensor(output_tensor);
+}
+
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp:167]: If this is used only in reduce, what about adding it on reduction_onednn.cpp instead?

File: src/plugins/intel_gpu/tests/test_cases/reduce_gpu_test.cpp [hunk 10]
```diff
@@ -1790,3 +1790,212 @@ INSTANTIATE_TEST_SUITE_P(reduce_gpu_b_fs_yx_fsv16_xy_i8,
...
+                            TestParamType_general_reduce_gpu(16, 16, 1, 1, 4, 11, format::bfyx, reduce_mode::mean, {reduce::along_y, reduce::along_f}, "reduce_ref", true, data_types::f32, false, data_types::f32),
+                            TestParamType_general_reduce_gpu(2, 16, 1, 1, 3, 6, format::bfyx, reduce_mode::mean, {reduce::along_f, reduce::along_y}, "reduce_ref", true, data_types::f32, false, data_types::f32),
+                            TestParamType_general_reduce_gpu(7, 4, 1, 1, 3, 6, format::bfyx, reduce_mode::mean, {reduce::along_x}, "reduce_ref", true, data_types::f32, false, data_types::f32),
+                            TestParamType_general_reduce_gpu(7, 32, 1, 1, 3, 6, format::bfyx, reduce_mode::mean, {reduce::along_b}, "reduce_ref", true, data_types::f32, false, data_types::f32),
+                            TestParamType_general_reduce_gpu(17, 4, 1, 1, 12, 15, format::bfyx, reduce_mode::mean, {reduce::along_b}, "reduce_ref", true, data_types::f32, false, data_types::f32)
+                        ), general_reduce_gpu::PrintToStringParamName);
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/test_cases/reduce_gpu_test.cpp:2000]: * I think fp16 is more important than fp32. But this test is all about fp32...
* Can we reduce the number of cases here?
* Is it inevitable to introduce new `ReduceOnednnTestBase` class? Can we just reuse `ReduceTestBase`?

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp (hunk 9 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp:167]: Moved.

File: src/plugins/intel_gpu/src/graph/impls/onednn/reduction_onednn.cpp (hunk 8 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/impls/onednn/reduction_onednn.cpp:38]: The loop is to check all axes whether the output size is matching to input size by reduced axes.
comments are added.

File: src/common/transformations/src/transformations/op_conversions/convert_reduce_to_reshape.cpp (hunk 7 above)
Comment (byungilm) [reply src/common/transformations/src/transformations/op_conversions/convert_reduce_to_reshape.cpp:79]: Added

File: src/plugins/intel_gpu/src/plugin/transformations_pipeline.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (byungilm) [reply src/plugins/intel_gpu/src/plugin/transformations_pipeline.cpp:202]: Fixed

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 11]
```diff
@@ -1833,6 +1831,23 @@ format layout_optimizer::get_preferred_format(program_node& node) {
                 expected = format::b_fs_yx_fsv32;
             }
         }
+    } else if (node.is_type<reduce>()) {
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:1834]: I think we need to check `use_onednn_impls`, too

Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:1834]: Applied

**Review by isanghao — APPROVED** [review]
LGTM, let me check the performance
